src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,508 @@
+
+    private readonly CancellationTokenSource _shutdownSource;
+    private volatile bool _disposed;
+
+    // The key for this dictionary is the Timestamp for the bucket, rounded down to the nearest RollupInSeconds... so it
+    // aggregates all of the metrics data for a particular time period. The Value is a dictionary for the metrics,
+    // each of which has a key that uniquely identifies it within the time period
+    internal ConcurrentDictionary<long, ConcurrentDictionary<string, Metric>> Buckets => _buckets.Value;
+
+    private readonly Lazy<ConcurrentDictionary<long, ConcurrentDictionary<string, Metric>>> _buckets
+        = new(() => new ConcurrentDictionary<long, ConcurrentDictionary<string, Metric>>());
+
+    private long _lastClearedStaleLocations = DateTimeOffset.UtcNow.GetDayBucketKey();
+    private readonly ConcurrentDictionary<long, HashSet<MetricResourceIdentifier>> _seenLocations = new();
+    private Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>> _pendingLocations = new();
+
+    private Task LoopTask { get; }
+
+    /// <summary>
+    /// MetricAggregator constructor.
+    /// </summary>
+    /// <param name="options">The <see cref="SentryOptions"/></param>
+    /// <param name="captureMetrics">The callback to be called to transmit aggregated metrics</param>
+    /// <param name="captureCodeLocations">The callback to be called to transmit new code locations</param>
+    /// <param name="shutdownSource">A <see cref="CancellationTokenSource"/></param>
+    /// <param name="disableLoopTask">
+    /// A boolean value indicating whether the Loop to flush metrics should run, for testing only.
+    /// </param>
+    /// <param name="flushInterval">An optional flushInterval, for testing only</param>
+    public MetricAggregator(SentryOptions options, Action<IEnumerable<Metric>> captureMetrics,
+        Action<CodeLocations> captureCodeLocations, CancellationTokenSource? shutdownSource = null,
+        bool disableLoopTask = false, TimeSpan? flushInterval = null)
+    {
+        _options = options;
+        _captureMetrics = captureMetrics;
+        _captureCodeLocations = captureCodeLocations;
+        _shutdownSource = shutdownSource ?? new CancellationTokenSource();
+        _flushInterval = flushInterval ?? TimeSpan.FromSeconds(5);
+
+        if (disableLoopTask)
+        {
+            // We can stop the loop from running during testing
+            _options.LogDebug("LoopTask disabled.");
+            LoopTask = Task.CompletedTask;
+        }
+        else
+        {
+            options.LogDebug("Starting MetricsAggregator.");
+            LoopTask = Task.Run(RunLoopAsync);
+        }
+    }
+
+    internal static string GetMetricBucketKey(MetricType type, string metricKey, MeasurementUnit unit,
+        IDictionary<string, string>? tags)
+    {
+        var typePrefix = type.ToStatsdType();
+        var serializedTags = GetTagsKey(tags);
```

Comment:
That's just mimicing the Python implementation. It might be more computationally expensive to generate a hash than to serialize these... although it would be less expensive in terms of memory footprint, so could go either way on this.